contracts/gateway/GatewayBase.sol

```diff
@@ -0,0 +1,395 @@
+     *  call of proveGateway method for same block height
+     */
+    event GatewayProven(
+        address _gateway,
+        uint256 _blockHeight,
+        bytes32 _storageRoot,
+        bool _wasAlreadyProved
+    );
+
+
+    /* Struct */
+
+    /**
+     * ActiveProcess stores the information related to in progress process
+     * like stake/mint unstake/redeem.
+     */
+    struct ActiveProcess {
+
+        /** latest message hash. */
+        bytes32 messageHash;
+
+        /** Outbox or Inbox process. */
+        MessageBus.MessageBoxType messageBoxType;
+    }
+
+
+    bytes32 constant STAKE_TYPEHASH = keccak256(
+        abi.encode(
+            "Stake(uint256 amount,address beneficiary,MessageBus.Message message)"
+        )
+    );
+
+    bytes32 constant REDEEM_TYPEHASH = keccak256(
+        abi.encode(
+            "Redeem(uint256 amount,address beneficiary,MessageBus.Message message)"
+        )
+    );
+    bytes32 constant GATEWAY_LINK_TYPEHASH = keccak256(
+        abi.encode(
+            "GatewayLink(bytes32 messageHash,MessageBus.Message message)"
+        )
+    );
+
+    /* constants */
+    /** Position of message bus in the storage */
+    uint8 constant MESSAGE_BOX_OFFSET = 1;
+    /** Penalty in bounty amount percentage charged to staker on revert staking
+     */
+    uint8 constant REVOCATION_PENALTY = 150;
+
+    /** Specifies if the Gateway and CoGateway contracts are linked. */
+    bool public linked;
+
+    /**
+     * Message box.
+     * @dev keep this is at location 1, in case this is changed then update
+     *      constant OUTBOX_OFFSET accordingly.
+     */
+    MessageBus.MessageBox messageBox;
+
+
+    /** Specifies if the Gateway is deactivated for any new staking process. */
+    bool public deactivated;
+
+    /** Organisation address. */
+    address public organisation;
+
+    /** amount of ERC20 which is staked by facilitator. */
+    uint256 public bounty;
+
+
+    /** address of core contract. */
+    CoreInterface public core;
+
+    /** path to prove merkle account proof for CoGateway contract. */
+    bytes internal encodedGatewayPath;
+
+    /** Remote gateway contract address. */
+    address public remoteGateway;
+
+    /** Gateway link message hash. */
+    bytes32 public gatewayLinkHash;
+
+    /** Maps messageHash to the Message object. */
+    mapping(bytes32 /*messageHash*/ => MessageBus.Message) messages;
+
+    /** Maps blockHeight to storageRoot*/
+    mapping(uint256 /* block height */ => bytes32) internal storageRoots;
+
+    /**
+     * Maps address to ActiveProcess object.
+     *
+     * Once the minting or redeem process is started the corresponding
+     * message hash is stored in ActiveProcess against the staker/redeemer
+     * address. This is used to restrict simultaneous/multiple minting and
+     * redeem for a particular address. This is also used to determine the
+     * nonce of the particular address. Refer getNonce for the details.
+     */
```

Comment:
As its now in base, update the comment to be more generic.

Reply:
Done